traffic_ops/traffic_ops_golang/login/login.go

```diff
@@ -104,3 +108,119 @@ func LoginHandler(db *sqlx.DB, cfg config.Config) http.HandlerFunc {
 		fmt.Fprintf(w, "%s", respBts)
 	}
 }
+
+func OauthLoginHandler(db *sqlx.DB, cfg config.Config) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		handleErrs := tc.GetHandleErrorsFunc(w, r)
+		defer r.Body.Close()
+		authenticated := false
+		resp := struct {
+			tc.Alerts
+		}{}
+
+		form := auth.PasswordForm{}
+		tokenForm := struct {
+			Token string `json:"t"`
+		}{}
+
+		if err := json.NewDecoder(r.Body).Decode(&tokenForm); err != nil {
+			handleErrs(http.StatusBadRequest, err)
+			return
+		}
+
+		encodedToken := tokenForm.Token
+
+		if encodedToken == "" {
+			log.Errorf("Token not found in request but is required")
+			handleErrs(http.StatusBadRequest, errors.New("Token not found in request but is required"))
+			return
+		}
+
+		decodedToken, err := jwt.Parse(encodedToken, func(unverifiedToken *jwt.Token) (interface{}, error) {
+			publicKeyUrl := unverifiedToken.Header["jku"].(string)
+			publicKeyId := unverifiedToken.Header["kid"].(string)
+
+			if !VerifyUrlOnWhiteList(publicKeyUrl, cfg.ConfigTrafficOpsGolang.WhitelistedOAuthUrls) {
+				return nil, errors.New("Key URL from token is not included in the whitelisted urls. Received: " + publicKeyUrl)
+			}
+
+			keys, err := jwk.FetchHTTP(publicKeyUrl)
+			if err != nil {
+				return nil, err
+			}
+
+			keyById := keys.LookupKeyID(publicKeyId)
+			selectedKey, err := keyById[0].Materialize()
+
+			if err != nil {
+				return nil, err
+			}
+
+			return selectedKey, nil
+		})
+		if err != nil {
+			handleErrs(http.StatusInternalServerError, errors.New("Error decoding token with message: "+err.Error()))
+			log.Errorf("Error decoding token: %s\n", err.Error())
+			return
+		}
+
+		authenticated = decodedToken.Valid
+
+		userId := decodedToken.Claims.(jwt.MapClaims)["sub"].(string)
+		form.Username = userId
+
+		userAllowed, err, blockingErr := auth.CheckLocalUserIsAllowed(form, db, time.Duration(cfg.DBQueryTimeoutSeconds)*time.Second)
+		if blockingErr != nil {
+			api.HandleErr(w, r, nil, http.StatusServiceUnavailable, nil, fmt.Errorf("error checking local user password: %s\n", blockingErr.Error()))
+			return
+		}
+		if err != nil {
+			log.Errorf("checking local user: %s\n", err.Error())
+			return
```

Comment:
For this one, I followed what is done in the original LoginHandler.  If this error is hit, then it logs the error and returns an "Invalid username or password." message because it failed to authorize and authenticate the user.  i think a call to api.HandleErr would be confusing because it wouldnt display the message that the user was invalid

Reply:
Alright :+1: sounds good

Reply:
This has to return an error to the client. Right now, by returning, this is just dropping the connection to the client, with no return code or body. If it was doing that before, that was a bug.

That's what `handleErrs` and `api.HandleErr` do. You could manually `w.WriteHeader(http.StatusInternalServerError)` here, but it needs to return _something_ to the client.

Reply:
So I've just confirmed - if you login with an OAuth SSO user that doesn't exist in the database, this will return a 200 response. In Traffic Portal, this results in it trying to load the dashboard (or redirect page, I think) but then failing to load resources with 401 errors and the user is redirected to the login page and asked to log in again.

The proper behavior here is to return a 403 with some appropriate message, and having looked at the normal user login code I don't see it bailing without setting an error response in this case. It _does_ do the error log the same way as this, but does not immediately return afterward, which is the problem here.

Reply:
done

Reply:
... is it though? At least here on GitHub I'm still seeing it just return after the `log.Errorf` call. I've refreshed a few times to be sure I was looking at the right thing.

Reply:
deleted that return.  now it will return a 403 with "Invalid user" if the user is either not in the db or doesnt have permission